### PR TITLE
Update code to work with new xcube type specifiers

### PR DESCRIPTION
**Pull request by pont-us**

### CODE REVIEW

File: xcube_cds/store.py [hunk 1]
```diff
@@ -38,11 +38,11 @@
 import xarray as xr
 
 import xcube.core.normalize
-from xcube.core.store import DataDescriptor
+from xcube.core.store import DataDescriptor, TypeSpecifier
```
Comment (TonioF) [inline xcube_cds/store.py:41]: Split this into two lines

File: xcube_cds/store.py [hunk 2]
```diff
@@ -478,7 +478,8 @@ def _create_time_range(t_start: str, t_end: str, t_interval: str):
         dt_start = dateutil.parser.isoparse(t_start)
         dt_end = datetime.datetime.now() if t_end is None \
             else dateutil.parser.isoparse(t_end)
-        period_number, period_unit = CDSDataOpener._parse_time_period(t_interval)
+        period_number, period_unit = \
+            CDSDataOpener._parse_time_period(t_interval)
```
Comment (TonioF) [inline xcube_cds/store.py:482]: Why split this into two lines?

File: xcube_cds/store.py [hunk 3]
```diff
@@ -749,11 +764,26 @@ def open_data(self, data_id: str, opener_id: Optional[str] = None,
...
+        if not CDSDataStore._is_type_specifier_compatible(type_specifier):
             raise DataStoreError(
-                f'Data type identifier must be "{TYPE_ID_DATASET}", '
-                f'but got "{type_id}"')
+                f'Supplied type specifier "{type_specifier}" is not compatible '
+                f'with "{TYPE_SPECIFIER_CUBE}."'
+            )
+
+    @staticmethod
+    def _is_type_specifier_compatible(
```
Comment (TonioF) [inline xcube_cds/store.py:775]: Even though this is just an internal method, I think it would be better to change its name to satisfies_type_specifier (or similar)

File: xcube_cds/store.py [hunk 4]
```diff
@@ -699,40 +700,54 @@ def get_data_store_params_schema(cls) -> JsonObjectSchema:
...
         raise NotImplementedError()
 
     def get_data_opener_ids(self, data_id: Optional[str] = None,
-                            type_id: Optional[str] = None) -> \
-            Tuple[str, ...]:
-        self._assert_valid_type_id(type_id)
+                            type_specifier: Optional[str] = None)\
+            -> Tuple[str, ...]:
+        self._validate_type_specifier(type_specifier)
         self._assert_valid_opener_id(data_id)
```
Comment (TonioF) [inline xcube_cds/store.py:745]: This is not part of the pull request, but this line does not make sense. I guess it should be self._validate_data_id(data_id) .

**Review by TonioF — CHANGES_REQUESTED** [review]
Looks mostly good, but please consider not separating imports by commas but instead using one import per line. Also, there were a lot of line breaks which did not seem necessary to me.

File: xcube_cds/store.py (hunk 1 above)
Comment (pont-us) [reply xcube_cds/store.py:41]: Done.

File: xcube_cds/store.py (hunk 2 above)
Comment (pont-us) [reply xcube_cds/store.py:482]: I try to keep xcube-cds code style compliant with the [xcube developer guide](https://xcube.readthedocs.io/en/latest/devguide.html#coding-style), which recommends PEP 8 style, which recommends [a maximum line length of 79 characters](https://www.python.org/dev/peps/pep-0008/#maximum-line-length).

**Comment by pont-us** [comment]
> Looks mostly good, but please consider not separating imports by commas but instead using one import per line.

Fixed throughout.

> Also, there were a lot of line breaks which did not seem necessary to me.

I try to keep xcube-cds code style compliant with the [xcube developer guide](https://xcube.readthedocs.io/en/latest/devguide.html#coding-style), which recommends PEP 8 style, which recommends [a maximum line length of 79 characters](https://www.python.org/dev/peps/pep-0008/#maximum-line-length). I break lines only when this limit is exceeded.

File: xcube_cds/store.py (hunk 2 above)
Comment (pont-us) [reply xcube_cds/store.py:482]: See [this reply](https://github.com/dcs4cop/xcube-cds/pull/21#issuecomment-724790253).

File: xcube_cds/store.py (hunk 3 above)
Comment (pont-us) [reply xcube_cds/store.py:775]: Agreed. I've changed it to `_is_type_specifier_satisfied`.

File: xcube_cds/store.py (hunk 4 above)
Comment (pont-us) [reply xcube_cds/store.py:745]: Thanks, fixed.

**Comment by pont-us** [comment]
Note that Travis builds will fail until https://github.com/dcs4cop/xcube/pull/359 is merged, since the xcube-cds Travis config tests with the head of the xcube master branch.

File: CHANGES.md [hunk 5]
```diff
@@ -8,6 +8,10 @@
    data store and data opener constructors
  - Update and improve the Jupyter demo notebooks
  - Various bug fixes
+ - Use the new xcube 0.5.2 data type specifier system
```
Comment (forman) [inline CHANGES.md:11]:
```suggestion
 - Use the new xcube 0.6 data type specifier system
```

File: environment.yml [hunk 6]
```diff
@@ -9,4 +9,4 @@ dependencies:
   - numpy >=1.17
   - python-dateutil >=2.8.1
   - xarray >=0.14.1
-  - xcube >=0.5.0
+  - xcube >=0.5.1  # TODO: update to xcube 0.5.2 as soon as it's available
```
Comment (forman) [inline environment.yml:12]:
```suggestion
  - xcube >=0.5.1  # TODO: update to xcube 0.6 as soon as it's available
```

File: xcube_cds/datasets/reanalysis_era5.py [hunk 7]
```diff
@@ -23,10 +23,12 @@
 import json
 import os
 import pathlib
-from typing import List, Optional, Dict, Tuple
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Tuple
```
Comment (forman) [inline xcube_cds/datasets/reanalysis_era5.py:29]: Its absolutely fine, but in xcube code we single component imports only for our own components.

File: xcube_cds/store.py [hunk 8]
```diff
@@ -749,11 +772,26 @@ def open_data(self, data_id: str, opener_id: Optional[str] = None,
...
-                f'Data type identifier must be "{TYPE_ID_DATASET}", '
-                f'but got "{type_id}"')
+                f'Supplied type specifier "{type_specifier}" is not compatible '
+                f'with "{TYPE_SPECIFIER_CUBE}."'
+            )
+
+    @staticmethod
+    def _is_type_specifier_satisfied(
+            type_specifier: Union[str, TypeSpecifier]
+            ) -> bool:
```
Comment (forman) [inline xcube_cds/store.py:785]: Weird function head formatting:

```suggestion
    def _is_type_specifier_satisfied(type_specifier: Union[str, TypeSpecifier]) -> bool:
```

xcube style:

If there is enough space, put all params in one line.
If not,
- first param directly behind first paren,
- next params aligned below first
- last paren directly behind last param

File: xcube_cds/store.py [hunk 9]
```diff
@@ -749,11 +772,26 @@ def open_data(self, data_id: str, opener_id: Optional[str] = None,
...
+        # At present, all datasets are available as cubes, so we simply check
+        # against TYPE_SPECIFIER_CUBE. If more (non-cube) datasets are added,
+        # the logic will have to be delegated to CDSDatasetHandler
+        # implementations.
+        if type_specifier is None:
+            return True
+        else:
+            type_specifier_cds = TypeSpecifier.normalize(TYPE_SPECIFIER_CUBE)
+            return type_specifier_cds.satisfies(type_specifier)
```
Comment (forman) [inline xcube_cds/store.py:794]:
```suggestion
        return TYPE_SPECIFIER_CUBE.satisfies(type_specifier)
```

**Review by forman — CHANGES_REQUESTED** [review]
Good, only few minor suggestions!

**Comment by TonioF** [comment]
> I try to keep xcube-cds code style compliant with the [xcube developer guide](https://xcube.readthedocs.io/en/latest/devguide.html#coding-style), which recommends PEP 8 style, which recommends [a maximum line length of 79 characters](https://www.python.org/dev/peps/pep-0008/#maximum-line-length). I break lines only when this limit is exceeded.

Interesting. I think I got fooled by the PyCharm messag that line breaks should occur after 120 characters here ( see here: https://stackoverflow.com/questions/26808681/why-does-pycharm-use-120-character-lines-even-though-pep8-specifies-79 ) . I'd love if we could agree on 99 characters for code, as PEP8 allows at max.

File: CHANGES.md (hunk 5 above)
Comment (pont-us) [reply CHANGES.md:11]: Done.

File: xcube_cds/datasets/reanalysis_era5.py (hunk 7 above)
Comment (pont-us) [reply xcube_cds/datasets/reanalysis_era5.py:29]: If it's fine, I'll stick with this "single import for everything" convention for the CDS Store.

File: xcube_cds/store.py (hunk 8 above)
Comment (pont-us) [reply xcube_cds/store.py:785]: All parameters (and type specifiers) on one line makes for 88 characters, exceeding PEP-8's 79-column limit. Even with the return type on the next line, it goes over. And when there's only one parameter, the three bullet points simplify to "all params on one line". But I can at least remove the line break between the parameter and the closing parenthesis.

**Comment by pont-us** [comment]
> I'd love if we could agree on 99 characters for code, as PEP8 allows at max.

Personally I'm an 80-column fundamentalist (in part because I love having the horizontal space for a side-by-side diff and some additional IDE panels), but if there's a democratic majority for 99 columns and if we officially mandate it in the xcube developer guide, I will comply :).
